### PR TITLE
Validate Provider and Quit on incorrect Specification File path

### PR DESCRIPTION
**Pull request by arya-hemanshu**
### Description
In case if we provide incorrect provider label the recipe runs but doesn't produce any result, this PR validates the provider name and if it is incorrect, the program quits with error and the name of the provider which is invalid

Incase wrong path of the specification file is passed the system continues of run rather than quitting , this PR also addresses that issue, on passing the incorrect path for the specification file, the program quits

#391 #325 

### Checklist
None

### CODE REVIEW

File: src/main/java/uk/org/tombolo/importer/utils/JSONReader.java [hunk 1]
```diff
@@ -192,6 +191,7 @@ public void setPrimaryNode(String primaryNode) {
         JsonReader reader = Json.createReader(is);
         JsonValue value = reader.read();
         convertTreeToHashMap(value, EMPTY_KEY);
+        getFlatJsonTree().add(individualSectionOfTree);
```
Comment (lorenaqendro) [inline src/main/java/uk/org/tombolo/importer/utils/JSONReader.java]: Why this change?

File: src/main/java/uk/org/tombolo/execution/DataExportEngine.java [hunk 2]
```diff
@@ -101,4 +103,38 @@ private void importDatasource(ImporterMatcher forceImports, DatasourceRecipe dat
...
+		Config importerConfiguration = null;
+		if (configFile != null && !"".equals(configFile)) {
+			importerConfiguration = ConfigUtils.loadConfig(
+					AbstractRunner.loadProperties("Configuration file", configFile));
+
+		}
+		return (Importer) Class.forName(importerClass).getDeclaredConstructor(Config.class).newInstance(importerConfiguration);
+	}
+
+	private String verifyProvider() throws Exception {
```
Comment (lorenaqendro) [inline src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: You might consider adding this method and/or check in  DataExportRunner.java where the other validators are.

File: src/main/java/uk/org/tombolo/execution/DataExportEngine.java [hunk 3]
```diff
@@ -101,4 +103,38 @@ private void importDatasource(ImporterMatcher forceImports, DatasourceRecipe dat
...
+		String validProvider = null;
+
+		File specificationFile = new File(DataExportRunner.executionSpecPath);
+		if (!specificationFile.exists()) return validProvider;
```
Comment (lorenaqendro) [inline src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: This check is already done in validateSpecification() that is called before the execute.

File: src/main/java/uk/org/tombolo/execution/DataExportEngine.java [hunk 4]
```diff
@@ -101,4 +103,38 @@ private void importDatasource(ImporterMatcher forceImports, DatasourceRecipe dat
...
+		JSONReader reader = new JSONReader(specificationFile, new ArrayList<>(Arrays.asList("importerClass", "provider")));
+		reader.getData();
+		ArrayList<String> importers = reader.getTagValueFromAllSections("importerClass");
```
Comment (lorenaqendro) [inline src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: Would this be still valid with the changes like in #296? In this case we could have the provider of a subject implicitly imported inside other importers.

File: src/main/java/uk/org/tombolo/execution/DataExportEngine.java [hunk 5]
```diff
@@ -41,6 +42,13 @@ public void execute(DataExportRecipe dataExportSpec, Writer writer) throws Excep
 	}
 
 	public void execute(DataExportRecipe dataExportSpec, Writer writer, ImporterMatcher forceImports) throws Exception {
+
+		//Check for Valid Providers
+		String vProvider = verifyProvider();
+		if (null != vProvider) {
```
Comment (lorenaqendro) [inline src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: typo != -> ==

File: src/main/java/uk/org/tombolo/importer/utils/JSONReader.java (hunk 1 above)
Comment (arya-hemanshu) [reply src/main/java/uk/org/tombolo/importer/utils/JSONReader.java]: because if you follow the logic, of how it gets and adds the value in final flat structure. The program return without adding the last section of the json file because of line 250, thus to add the final section of the json file, this step is necessary, it was bug in previous implementation which i realised while doing that issue

File: src/main/java/uk/org/tombolo/execution/DataExportEngine.java (hunk 3 above)
Comment (arya-hemanshu) [reply src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: This is necessary to see if the method is invoked by tests or actual recipe file. As tests wont have a file but would execute this method

File: src/main/java/uk/org/tombolo/execution/DataExportEngine.java (hunk 5 above)
Comment (arya-hemanshu) [reply src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: No, this code is actually checking for when provider is not null, if it is null then it mean all providers in recipe are valid and this method is not invoked by test cases, please follow the verifyProvider method again, in case you still find it ambiguous

File: src/main/java/uk/org/tombolo/execution/DataExportEngine.java (hunk 4 above)
Comment (arya-hemanshu) [reply src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: Yes it checks for all the providers, irrespective of the level at which they are provided in the recipe, all the providers are checked in a recipe for their validity

File: src/main/java/uk/org/tombolo/execution/DataExportEngine.java (hunk 2 above)
Comment (arya-hemanshu) [reply src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: There are 2 reasons why i didn't do it there,
1) DataExportRunner.java does the initial checks for arguments and structure of recipe, nothing related to importers or anything else. Thus i thought of leaving it the way it is, to have it confined till initial checks.
2) If we do that in DataExportRunner.java we have to repeat the initialiseImporters code and create a dummy variable for Config class, which i think is not good as creating dummy variables are usually an approach taken while creating test.

File: src/main/java/uk/org/tombolo/execution/DataExportEngine.java [hunk 6]
```diff
@@ -101,4 +103,38 @@ private void importDatasource(ImporterMatcher forceImports, DatasourceRecipe dat
...
+		JSONReader reader = new JSONReader(specificationFile, new ArrayList<>(Arrays.asList("importerClass", "provider")));
+		reader.getData();
+		ArrayList<String> importers = reader.getTagValueFromAllSections("importerClass");
+		ArrayList<String> providers = reader.getTagValueFromAllSections("provider");
+
+		mainLoop:
+		for (String importer : importers) {
+			for (String provider : providers) {
+				Importer imp = initialiseImporter(importer, "");
+				if (!provider.equals(imp.getProvider().getLabel())) {
```
Comment (lorenaqendro) [inline src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: Sorry I missed this failure in the first review:
This loop fails for me because it continues if it finds the right provider and fails the following iteration with a different provider for the same importer. I tested it with `src/main/resources/executions/examples/london-cycle-traffic-air-quality-lsoa.json`. Does it fail for you too?

File: src/main/java/uk/org/tombolo/execution/DataExportEngine.java (hunk 4 above)
Comment (lorenaqendro) [reply src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: This will miss the modelling fields providers, right?

Comment (arya-hemanshu) [reply src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: Yeah it will miss the modelling fields, it is only for main recipe

File: src/main/java/uk/org/tombolo/execution/DataExportEngine.java (hunk 6 above)
Comment (arya-hemanshu) [reply src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: good catch, yeah it does, because in fields it doesn't know which importer class to look for and falls over now this kind of makes it hard to fix

Comment (arya-hemanshu) [reply src/main/java/uk/org/tombolo/execution/DataExportEngine.java]: good catch, it does fail for me, i got what you mean. :)

**Comment by arya-hemanshu** [comment]
merged with #433

**Comment by arya-hemanshu** [comment]
merged with #433
